### ISSUE TITLE
Installation does not work on specific hardware

### ISSUE THREAD

**Issue by htejun** [iocost-benchmark/resctl-demo-image-recipe#35]
Installation stalls after image is written to target disk. It didn't drop to shell prompt. The console wasn't responsive to any input. Switched to the second console w/ ctrl-alt-f2 and collected the following logs. Will attach the screenshot too.

![Screenshot from 2024-01-17 16-34-51](https://github.com/iocost-benchmark/resctl-demo-image-recipe/assets/1051723/657c6def-e4da-404d-8777-cd3e19fa7ca9)

[logs.zip](https://github.com/iocost-benchmark/resctl-demo-image-recipe/files/13972440/logs.zip)

**Comment by santoshmahto7** [comment]
Thanks. It's very helpful to confirm from screenshot that mount was failing.
Not having bash prompt and screen install are expected since that's missing in code, I will add patch for that.

**Comment by santoshmahto7** [comment]
@htejun :  With latest PR (merged), now bash will be prompted when mount fails.  Could you again check in that bash prompt what has caused the mount failure. Thanks in advance.
I would be interested in these too things.
1.  Whether `/mnt/rootfs/boo/efi`  directory is present or not ?
     -  if not, then try to create it manually and check for any failure.
3.  If yes, then try to mount it manually and then exit bash to continue with installation and confirm if kexec works or not.

**Comment by htejun** [comment]
![Screenshot from 2024-01-21 15-21-07](https://github.com/iocost-benchmark/resctl-demo-image-recipe/assets/1051723/ad761e0d-1946-4141-b7e1-a2e2ef9389b1)
![Screenshot from 2024-01-21 15-22-17](https://github.com/iocost-benchmark/resctl-demo-image-recipe/assets/1051723/73e44771-a387-42f0-ba98-4eb48c581833)

The errors seem already visible in the previous capture. `/etc/fstab` doesn't have an entry for the boot mount and is it trying to mount `/dev/p1`?

**Comment by santoshmahto7** [comment]
IIUC Entry in "/etc/fstab" is not mandatory and only checked when target or source is not available. As you can see, manual mount from command line is successful. I need to debug why mount looks for entry in /etc/fstab.

Also It tries to mount /dev/nvme0n1p1 not /dev/p1. "/dev/p1" is printed as log to denote first partition, But updating it to actually print partition name will be better, will do that.

**Comment by htejun** [comment]
If you make it log the commands that it's running somewhere, would that help?

**Comment by santoshmahto7** [comment]
@htejun I created one PR https://github.com/iocost-benchmark/resctl-demo-image-recipe/pull/37 to fix this issue.

I was also able to reproduce issue on my machine (choose another machine) and investigated the issue. The reason of failure was strange although.
After copying (`bmaptool copy`) image to the target device, `lsblk -n -o PATH` was not listing the partitions of target device for some seconds, so there was a synchronisation issue.
I added some `sleep and udevadm --trigger settle` for flasher os to be ready enough to see the target partitions.

I am still doing more iterations to verify the issue fix, but It would be nice if you could also confirm on your side (after patch is merged)

**Comment by htejun** [comment]
Ah okay, you might need to trigger partition scan manually and wait for it. bmap is just overwriting the disk, right? I don't know what's triggering partition scan but something is doing belatedly and the script is likely not waiting for it. Can you replace the sleep and udevadm call with `blockdev --rereadpt` and see whether that works?

**Comment by santoshmahto7** [comment]
> Ah okay, you might need to trigger partition scan manually and wait for it. bmap is just overwriting the disk, right? I don't know what's triggering partition scan but something is doing belatedly and the script is likely not waiting for it. Can you replace the sleep and udevadm call with `blockdev --rereadpt` and see whether that works?

I will try that, and yes bmap is overwriting the disk (equivalent to dd) and script misses the updated  partitions list in target device.

**Comment by santoshmahto7** [comment]
@htejun : `blockdev --rereadpt` seems to works (tried 3 iterations) without any issue. 
I opened a PR here: https://github.com/iocost-benchmark/resctl-demo-image-recipe/pull/38
Could you also check on your side?

**Comment by htejun** [comment]
Will do. Thanks.

**Comment by santoshmahto7** [comment]
I created another PR https://github.com/iocost-benchmark/resctl-demo-image-recipe/pull/41/files
@htejun Once above PR is merged, It would be great to test and confirm if issue are resolved. Thanks

**Comment by obbardc** [comment]
Looks like the root cause for this is that images are built assuming the disk has 512 byte sectors, but the target hardware has 4096 byte sectors.

It can be reproduced in qemu by creating a disk with 4096 byte sectors (I will add a patch to add this case into the qemu scripts):
![image](https://github.com/user-attachments/assets/a395d555-8236-4355-9fb8-d36cb4258a8c)

**Comment by obbardc** [comment]
Fixed by #53
